File: rust/src/code/swim_in_rising_water.rs
use std::cmp::{max, Reverse};
use std::collections::{BinaryHeap, HashSet};

type X = usize;
type Y = usize;

struct Solution;

impl Solution {
    // NOTE: time complexity O((n^2) * log n)
    pub fn swim_in_water(grid: Vec<Vec<i32>>) -> i32 {
        let n: usize = grid.len();
        let mut min_heap: BinaryHeap<Reverse<(i32, X, Y)>> =
            BinaryHeap::from([Reverse((grid[0][0], 0, 0))]);
        let mut visited: HashSet<(X, Y)> = HashSet::from([(0, 0)]);
        let directions: Vec<(isize, isize)> = vec![(1, 0), (0, 1), (-1, 0), (0, -1)];

        while let Some(Reverse((h, x, y))) = min_heap.pop() {
            if x == n - 1 && y == n - 1 {
                return h;
            }

            for (dx, dy) in directions.iter() {
                let nx: isize = (x as isize + *dx) as isize;
                let ny: isize = (y as isize + *dy) as isize;
                if nx >= 0
                    && ny >= 0
                    && (nx as usize) < n
                    && (ny as usize) < n
                    && !visited.contains(&(nx as usize, ny as usize))
                {
                    let nx = nx as usize;
                    let ny = ny as usize;
                    visited.insert((nx, ny));
                    min_heap.push(Reverse((max(grid[ny][nx], h), nx, ny)));
                } else {
                    continue;
                }
            }
        }

        unreachable!("should not reach here");
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_swim_in_water_case_1() {
        // arrange
        let grid = vec![vec![0, 2], vec![1, 3]];
        let expected = 3;

        // act
        let actual = Solution::swim_in_water(grid);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_swim_in_water_case_2() {
        // arrange
        let grid = vec![
            vec![0, 1, 2, 3, 4],
            vec![24, 23, 22, 21, 5],
            vec![12, 13, 14, 15, 16],
            vec![11, 17, 18, 19, 20],
            vec![10, 9, 8, 7, 6],
        ];
        let expected = 16;

        // act
        let actual = Solution::swim_in_water(grid);

        // assert
        assert_eq!(actual, expected);
    }
}

File: python/swim_in_rising_water.py
import heapq


class Solution:
    def swimInWater(self, grid: list[list[int]]) -> int:
        """time complexity: O(n^2 * log n)"""
        n: int = len(grid)
        minHeap: list[tuple[int, int, int]] = [(grid[0][0], 0, 0)]
        visited: set[tuple[int, int]] = set([(0, 0)])
        direction: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        while len(minHeap) > 0:
            h, x, y = heapq.heappop(minHeap)
            if x == n - 1 and y == n - 1:
                return h
            for dx, dy in direction:
                nx: int = x + dx
                ny: int = y + dy
                if (
                    nx >= 0
                    and ny >= 0
                    and nx < n
                    and ny < n
                    and (nx, ny) not in visited
                ):
                    heapq.heappush(minHeap, (max(h, grid[ny][nx]), nx, ny))
                    visited.add((nx, ny))

        raise ValueError("Unreachable")


def test_swimInWater_case_1():
    # arrange
    grid: list[list[int]] = [[0, 2], [1, 3]]
    expected: int = 3

    # act
    actual = Solution().swimInWater(grid)

    # assert
    assert actual == expected


def test_swimInWater_case_2():
    # arrange
    grid: list[list[int]] = [
        [0, 1, 2, 3, 4],
        [24, 23, 22, 21, 5],
        [12, 13, 14, 15, 16],
        [11, 17, 18, 19, 20],
        [10, 9, 8, 7, 6],
    ]
    expected: int = 16

    # act
    actual = Solution().swimInWater(grid)

    # assert
    assert actual == expected
